capsules/extra/src/nonvolatile_storage_driver.rs

```diff
@@ -202,7 +464,427 @@ impl<'a> NonvolatileStorage<'a> {
+                    {
+                        return Err(ErrorCode::FAIL);
+                    }
+
+                    let Some(header) = AppRegionHeader::new(region.version, shortid, region.length) else {
+                        return Err(ErrorCode::FAIL);
+                    };
+
+                    // write this new region header to the end of the existing ones
+                    self.write_region_header(processid, &region, &header, new_header_addr)
+                } else {
+                    // this app never requested to be allocated or its
+                    // region was already allocated
+                    Ok(())
+                }
+            })
+            .unwrap_or(Err(ErrorCode::FAIL))
+    }
+
+    // Read the header of an app's storage region. The region_header_address argument
+    // describes the start of the **header** and not the usable region itself.
+    fn read_region_header(&self, region_header_address: usize) -> Result<(), ErrorCode> {
+        if DEBUG {
+            debug!(
+                "[NONVOLATILE_STORAGE_DRIVER]: Reading region header from {:#x}",
+                region_header_address
+            );
+        }
+        self.enqueue_command(
+            NonvolatileCommand::HeaderRead(region_header_address),
+            region_header_address,
+            REGION_HEADER_LEN,
+            None,
+        )
+    }
+
+    fn write_region_header(
+        &self,
+        processid: ProcessId,
+        region: &AppRegion,
+        region_header: &AppRegionHeader,
+        region_header_address: usize,
+    ) -> Result<(), ErrorCode> {
+        if DEBUG {
+            debug!(
+                "[NONVOLATILE_STORAGE_DRIVER]: Writing region header to {:#x}. Header: {:#x?}",
+                region_header_address, region_header
+            );
+        }
+
+        let header_slice = region_header.to_bytes();
+
+        self.header_buffer.map_or(Err(ErrorCode::NOMEM), |buffer| {
+            for (i, c) in buffer
+                .iter_mut()
+                .enumerate()
+            {
+                *c = header_slice[i];
+            }
+
+            Ok(())
+        })?;
+
+        self.enqueue_command(
+            NonvolatileCommand::HeaderWrite(processid, *region),
+            region_header_address,
+            REGION_HEADER_LEN,
+            None,
+        )
+    }
+
+    fn erase_region_content(&self, processid: ProcessId, region: AppRegion) -> Result<(), ErrorCode> {
+        if DEBUG {
+            debug!(
+                "[NONVOLATILE_STORAGE_DRIVER]: Erasing content of region with header starting at {:#x}",
+               region.offset 
+            );
+        }
+        self.region_erase_buffer.map_or(Err(ErrorCode::NOMEM), |buffer| {
+            // clear the erase buffer in case there was any nonzero value there
+            for c in buffer.iter_mut() {
+                *c = 0;
+            }
+            Ok(())
+        })?;
```

Comment:
In future we should add a `clear()` function to the storage HIL so we can just call that, but this is ok for now